exporter/clickhouseexporter/internal/sum_metrics.go

```diff
@@ -0,0 +1,171 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/clickhouseexporter/internal"
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+const (
+	// language=ClickHouse SQL
+	createSumTableSQL = `
+CREATE TABLE IF NOT EXISTS %s_sum (
+    ResourceAttributes Map(LowCardinality(String), String) CODEC(ZSTD(1)),
+    ResourceSchemaUrl String CODEC(ZSTD(1)),
+    ScopeName String CODEC(ZSTD(1)),
+    ScopeVersion String CODEC(ZSTD(1)),
+    ScopeAttributes Map(LowCardinality(String), String) CODEC(ZSTD(1)),
+    ScopeDroppedAttrCount UInt32 CODEC(ZSTD(1)),
+    ScopeSchemaUrl String CODEC(ZSTD(1)),
+    MetricName String CODEC(ZSTD(1)),
+    MetricDescription String CODEC(ZSTD(1)),
+    MetricUnit String CODEC(ZSTD(1)),
+    Attributes Map(LowCardinality(String), String) CODEC(ZSTD(1)),
+	StartTimeUnix DateTime64(9) CODEC(Delta, ZSTD(1)),
+	TimeUnix DateTime64(9) CODEC(Delta, ZSTD(1)),
+	ValueAsDouble Float64 CODEC(ZSTD(1)),
+	ValueAsInt UInt32 CODEC(ZSTD(1)),
+	Flags UInt32  CODEC(ZSTD(1)),
+    Exemplars Nested (
+		FilteredAttributes Map(LowCardinality(String), String),
+		TimeUnix DateTime64(9),
+		ValueAsDouble Float64,
+		ValueAsInt Int64,
+		SpanId String,
+		TraceId String
+    ) CODEC(ZSTD(1)),
+    AggTemp Int32 CODEC(ZSTD(1)),
+	IsMonotonic Boolean CODEC(Delta, ZSTD(1))
+) ENGINE MergeTree()
+%s
+PARTITION BY toDate(TimeUnix)
+ORDER BY (MetricName, Attributes, toUnixTimestamp64Nano(TimeUnix))
+SETTINGS index_granularity=8192, ttl_only_drop_parts = 1;
+`
+	// language=ClickHouse SQL
+	insertSumTableSQL = `INSERT INTO %s_sum (
+    ResourceAttributes,
+    ResourceSchemaUrl,
+    ScopeName,
+    ScopeVersion,
+    ScopeAttributes,
+	ScopeDroppedAttrCount,
+    ScopeSchemaUrl,
+    MetricName,
+    MetricDescription,
+    MetricUnit,
+    Attributes,
+    TimeUnix,
+    ValueAsDouble,
+    ValueAsInt,
+    Flags,
+    Exemplars.FilteredAttributes,
+	Exemplars.TimeUnix,
+    Exemplars.ValueAsDouble,
+    Exemplars.ValueAsInt,
+    Exemplars.SpanId,
+    Exemplars.TraceId,
+	AggTemp,
+	IsMonotonic) VALUES `
+	sumPlaceholders = "(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"
+)
+
+type sumModel struct {
+	metricName        string
+	metricDescription string
+	metricUnit        string
+	sum               pmetric.Sum
+}
+
+type SumMetrics struct {
+	sumModel  []*sumModel
+	metadata  *MetricsMetaData
+	insertSQL string
+}
+
+func (s *SumMetrics) insert(ctx context.Context, tx *sql.Tx, logger *zap.Logger) error {
+	var valuePlaceholders []string
+	var valueArgs []interface{}
+
+	for _, model := range s.sumModel {
+		for i := 0; i < model.sum.DataPoints().Len(); i++ {
+			dp := model.sum.DataPoints().At(i)
+			valuePlaceholders = append(valuePlaceholders, sumPlaceholders)
```

Comment:
I'm not sure what this does?

Reply:
The `sumPlaceholders = "(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"` is a placeholder for insert string, if there are two metric entities, the insert query string will like this:
```
INSERT INTO otel_metrics_sum (
    ResourceAttributes,
    ResourceSchemaUrl,
    ScopeName,
    ScopeVersion,
    ScopeAttributes,
	ScopeDroppedAttrCount,
    ScopeSchemaUrl,
    MetricName,
    MetricDescription,
    MetricUnit,
    Attributes,
    TimeUnix,
    ValueAsDouble,
    ValueAsInt,
    Flags,
    Exemplars.FilteredAttributes,
	Exemplars.TimeUnix,
    Exemplars.ValueAsDouble,
    Exemplars.ValueAsInt,
    Exemplars.SpanId,
    Exemplars.TraceId,
	AggTemp,
	IsMonotonic) VALUES  (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?),(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
```

Reply:
Ok!